root/static/scripts/account/components/PreferencesForm.js

```diff
@@ -0,0 +1,257 @@
+import FormSubmit from '../../../../components/FormSubmit';
+import {l, N_l} from '../../common/i18n';
+import {Lens, prop, set, compose3} from '../../common/utility/lens';
+import hydrate from '../../../../utility/hydrate';
+
+type PreferencesFormT = FormT<{|
+  +datetime_format: FieldT<string>,
+  +email_on_no_vote: FieldT<boolean>,
+  +email_on_notes: FieldT<boolean>,
+  +email_on_vote: FieldT<boolean>,
+  +notify_via_email: FieldT<boolean>,
+  +public_ratings: FieldT<boolean>,
+  +public_subscriptions: FieldT<boolean>,
+  +public_tags: FieldT<boolean>,
+  +show_gravatar: FieldT<boolean>,
+  +subscribe_to_created_artists: FieldT<boolean>,
+  +subscribe_to_created_labels: FieldT<boolean>,
+  +subscribe_to_created_series: FieldT<boolean>,
+  +subscriptions_email_period: FieldT<string>,
+  +timezone: FieldT<string>,
+|}>;
+
+type Props = {|
+  +form: PreferencesFormT,
+  +timezone_options: MaybeGroupedOptionsT,
+|};
+
+type State = {|
+  form: PreferencesFormT,
+  timezoneOptions: MaybeGroupedOptionsT,
+|};
+
+const allowedDateTimeFormats = [
+  '%Y-%m-%d %H:%M %Z',
+  '%c',
+  '%x %X',
+  '%X %x',
+  '%A %B %e %Y, %H:%M',
+  '%d %B %Y %H:%M',
+  '%a %b %e %Y, %H:%M',
+  '%d %b %Y %H:%M',
+  '%d/%m/%Y %H:%M',
+  '%m/%d/%Y %H:%M',
+  '%d.%m.%Y %H:%M',
+  '%m.%d.%Y %H:%M',
+];
+
+function buildDateTimeFormatOptions(timezone) {
+  const hereAndNow = moment.tz(timezone);
+  return {
+    grouped: false,
+    options: allowedDateTimeFormats.map(a => ({
+      label: hereAndNow.strftime(a),
+      value: a,
+    })),
+  };
+}
+
+const subscriptionsEmailPeriodOptions = {
+  grouped: false,
+  options: [
+    {label: N_l('Daily'), value: 'daily'},
+    {label: N_l('Weekly'), value: 'weekly'},
+    {label: N_l('Never'), value: 'never'},
+  ],
+};
+
+const timezoneFieldLens: Lens<PreferencesFormT, string> =
+  compose3(prop('field'), prop('timezone'), prop('value'));
+
+const dateTimeFormatFieldLens: Lens<PreferencesFormT, string> =
+  compose3(prop('field'), prop('datetime_format'), prop('value'));
+
+const subscriptionsEmailPeriodFieldLens: Lens<PreferencesFormT, string> =
+  compose3(prop('field'), prop('subscriptions_email_period'), prop('value'));
+
+class PreferencesForm extends React.Component<Props, State> {
+  constructor(props: Props) {
+    super(props);
+    this.state = {form: props.form, timezoneOptions: props.timezone_options};
+    this.handleTimezoneChange = this.handleTimezoneChange.bind(this);
+    this.handleTimezoneGuess = this.handleTimezoneGuess.bind(this);
+    this.handleDateTimeFormatChange =
+      this.handleDateTimeFormatChange.bind(this);
+    this.handleSubscriptionsEmailPeriodChange =
+      this.handleSubscriptionsEmailPeriodChange.bind(this);
+  }
+
+  handleTimezoneChange: (e: SyntheticEvent<HTMLSelectElement>) => void;
+
+  handleTimezoneChange(e: SyntheticEvent<HTMLSelectElement>) {
+    const selectedTimezone = e.currentTarget.value;
+    this.setState(prevState => ({
+      form: set(
```

Comment:
As we convert more forms to React maybe we'll want to (later) write some helper utility to generate all these lenses/event handlers for us based on a concise description of the fields.